### PR TITLE
Add description of versioning and branch layout

### PR DESCRIPTION
**Pull request by Deamon5550**
A description of the changes to the repository to assist developers. Feedback appreciated.

### CODE REVIEW

**Comment by Spongy** [comment]
Setting up PR reference, please wait...

File: source/contributing/versioning.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+This change means that every time we're making a release which has non-backwards-compatible
+(breaking) changes we have to increment the major version of the api by one (eg. from 3.0.0 ->
+4.0.0). The layout of our branches which is described below are designed to assist with being able
+to make minor releases without a breaking change forcing us to make it a major release. This branch
+layout applies to the SpongeAPI, SpongeCommon, SpongeForge, and SpongeVanilla repositories.
+
+The Core Branches
+=================
+
+There are three branches which form the core of our repositories, they are master, bleeding, and
```
Comment (kashike) [inline source/contributing/versioning.rst]: The branch names should stand out - display as `master`, etc?

File: source/contributing/versioning.rst [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+================
+
+New features or changes will continue to be done in a feature/foo or fix/bar branch. When merging
+back into a development branch (master or bleeding) you should consider whether the changes are
+breaking or are strictly backwards-compatible. If the changes are purely new features or
+binary-compatible bugfixes then the feature branch can be merged into the master branch. If the
+changes include any breaking changes however then the feature branch must be merged into the
+bleeding branch to be included in the next major release.
+
+https://dl.dropboxusercontent.com/u/17223377/bin/Sponge/sponge-flow.png
```
Comment (kashike) [inline source/contributing/versioning.rst]: Should be an image in the docs repo, guaranteed not to break

Comment (Deamon5550) [inline source/contributing/versioning.rst]: yeah it was just temporarily included

**Comment by Inscrutable** [comment]
Remember that title underlines need to match the text length. Travis does ;)

**Comment by Tzky** [comment]
Q:
What happens if we need or want to hotfix an old major version? 

Example:
4.0.0 got tagged on the ``release`` branch and we need to provide a hotfix for 3.1.0 (so bump 3.1.0 -> 3.1.1). 3.1.1 should be on the release branch afterwards, but there's already 4.0.0.

I think that
* rewriting the commit history and rebasig 4.0.0 or
* having 3.1.1 after 4.0.0 on the timeline

isn't an option.

**Comment by gabizou** [comment]
There'd be a branch made for 3.x etc. I'd assume, because then backwards compatible patches could be made for previous versions, however it would mean that branch would permanently be out of date to release.

**Comment by Tzky** [comment]
We should probably split the page for Implementation+API and the Docs, as the Docs won't use the implementation versioning and branching system.

**Comment by Tzky** [comment]
@Deamon5550 
Are you intending to add the branching picture gabizou showed off in IRC a few days ago?

**Comment by Mumfrey** [comment]
I'm making a pretty version in visio. *(Please, try to withhold your surprise)*

**Comment by Tzky** [comment]
@mumfrey, maybe the style shown here (https://docs.spongepowered.org/en/contributing/howtogit.html) fits?

**Comment by Tzky** [comment]
Little reminder to myself:
``contributing/guidelines`` page needs a link to this PR.

**Comment by Tzky** [comment]
@Mumfrey:
We decided that we'll merge this PR for now. This leaves more time for you (doing your visio magic) and we have this merged for now.

**Pull request merged**